File: static/cpp/java/src/main/java/com/epi/SnakeString.java
package com.epi;

import java.util.Random;

public class SnakeString {
  // @include
  public static String snakeString(String s) {
    StringBuilder result = new StringBuilder();
    // Outputs the first row, i.e., s[1], s[5], s[9], ...
    for (int i = 1; i < s.length(); i += 4) {
      result.append(s.charAt(i));
    }
    // Outputs the second row, i.e., s[0], s[2], s[4], ...
    for (int i = 0; i < s.length(); i += 2) {
      result.append(s.charAt(i));
    }
    // Outputs the third row, i.e., s[3], s[7], s[11], ...
    for (int i = 3; i < s.length(); i += 4) {
      result.append(s.charAt(i));
    }
    return result.toString();
  }
  // @exclude

  private static String randString(int len) {
    Random r = new Random();
    StringBuilder ret = new StringBuilder(len);
    while (len-- > 0) {
      ret.append((char)(r.nextInt(26) + 'A'));
    }
    return ret.toString();
  }

  private static void smallTest() {
    assert(snakeString("Hello World!").equals("e lHloWrdlo!"));
  }

  public static void main(String[] args) {
    smallTest();
    Random r = new Random();
    String s;
    if (args.length == 1) {
      s = args[0];
    } else {
      s = randString(r.nextInt(100) + 1);
    }
    System.out.println(snakeString(s));
  }
}

File: static/python/snake_string.py
import sys
import random
import string


# @include
def snake_string(s):
    result = []
    # Outputs the first row, i.e., s[1], s[5], s[9], ...
    for i in range(1, len(s), 4):
        result.append(s[i])
    # Outputs the second row, i.e., s[0], s[2], s[4], ...
    for i in range(0, len(s), 2):
        result.append(s[i])
    # Outputs the third row, i.e., s[3], s[7], s[11], ...
    for i in range(3, len(s), 4):
        result.append(s[i])
    return ''.join(result)


# Python solution uses slicing with right steps.
def snake_string_pythonic(s):
    return s[1::4] + s[::2] + s[3::4]
# @exclude


def rand_string(length):
    return ''.join(random.choice(string.ascii_uppercase) for _ in range(length))


def small_test():
    assert snake_string(
        'Hello World!') == 'e lHloWrdlo!' == snake_string_pythonic(
            'Hello World!') == 'e lHloWrdlo!'


def main():
    small_test()
    s = sys.argv[1] if len(sys.argv) == 2 else rand_string(
        random.randint(1, 100))
    print(snake_string(s))
    assert snake_string(s) == snake_string_pythonic(s)


if __name__ == '__main__':
    main()
